python/perf-kernels/multreduce_matmul_kernel.py

```diff
@@ -1,45 +1,464 @@
     c_mask = (offs_cm[:, None] < M) & (offs_cn[None, :] < N)
     tl.store(c_ptrs, c, mask=c_mask)
+
+
+@triton.autotune(configs=get_triton_dot_autotune_configs(), key=get_triton_autotune_key())
+@triton.heuristics(get_triton_heuristics())
+@triton.jit
+def triton_dot_matmul_kernel(a_ptr, b_ptr, c_ptr, bias_ptr,  #
+                             M: int, N: int, K: int,  #
+                             stride_am: int, stride_ak: int,  #
+                             stride_bk: int, stride_bn: int,  #
+                             stride_cm: int, stride_cn: int,  #
+                             stride_bias: int,  #
+                             BLOCK_SIZE_M: tl.constexpr, BLOCK_SIZE_N: tl.constexpr, BLOCK_SIZE_K: tl.constexpr,  #
+                             USE_BIAS: tl.constexpr, EVEN_K: tl.constexpr  #
+                             ):
+    triton_matmul_kernel(a_ptr, b_ptr, c_ptr, bias_ptr,  #
+                         M, N, K,  #
+                         stride_am, stride_ak,  #
+                         stride_bk, stride_bn,  #
+                         stride_cm, stride_cn,  #
+                         stride_bias,  #
+                         BLOCK_SIZE_M=BLOCK_SIZE_M, BLOCK_SIZE_N=BLOCK_SIZE_N, BLOCK_SIZE_K=BLOCK_SIZE_K,  #
+                         USE_BIAS=USE_BIAS, USE_DOT=True, EVEN_K=EVEN_K)
+
+
+@triton.autotune(configs=get_triton_multreduce_autotune_configs(), key=get_triton_autotune_key())
+@triton.heuristics(get_triton_heuristics())
+@triton.jit
+def triton_multreduce_matmul_kernel(a_ptr, b_ptr, c_ptr, bias_ptr,  #
+                                    M: int, N: int, K: int,  #
+                                    stride_am: int, stride_ak: int,  #
+                                    stride_bk: int, stride_bn: int,  #
+                                    stride_cm: int, stride_cn: int,  #
+                                    stride_bias: int,  #
+                                    BLOCK_SIZE_M: tl.constexpr, BLOCK_SIZE_N: tl.constexpr,
+                                    BLOCK_SIZE_K: tl.constexpr,  #
+                                    USE_BIAS: tl.constexpr, EVEN_K: tl.constexpr  #
+                                    ):
+    triton_matmul_kernel(a_ptr, b_ptr, c_ptr, bias_ptr,  #
+                         M, N, K,  #
+                         stride_am, stride_ak,  #
+                         stride_bk, stride_bn,  #
+                         stride_cm, stride_cn,  #
+                         stride_bias,  #
+                         BLOCK_SIZE_M=BLOCK_SIZE_M, BLOCK_SIZE_N=BLOCK_SIZE_N, BLOCK_SIZE_K=BLOCK_SIZE_K,  #
+                         USE_BIAS=USE_BIAS, USE_DOT=False, EVEN_K=EVEN_K)
+
+
+def triton_matmul(triton_provider: str, a: Tensor, b: Tensor, bias: Optional[Tensor]) -> Tensor:
+    assert triton_provider in ["triton-dot", "triton-multreduce"]
+
+    M: int
+    N: int
+    K: int
+    M, K = a.shape
+    _, N = b.shape
+
+    c: Tensor = torch.empty((M, N), device=a.device, dtype=a.dtype)
```

Comment:
this should be torch.zeros instead of torch.empty ?

Reply:
Thanks for your comment Xiaohu.

I think we can use either `torch.empty` or `torch.zeros` here because the Triton kernel initializes the accumulator as `accumulator = tl.zeros((BLOCK_SIZE_M, BLOCK_SIZE_N), dtype=acc_dtype)`.

However, I can use `torch.zeros` instead of `torch.empty` if you think it's better.

Reply:
I think you will have correctness issue if your output is not initialised ?

Reply:
Resolved by https://github.com/ROCm/triton/pull/637/commits/02ddbfd90b6ba11b6f334a6a4f26d4dfa701a5b9.